### PR TITLE
store/tikv: support parallel && retry for batchGet.

### PR DESCRIPTION
**Pull request by disksing**
@coocood @zimulala @shenli

### CODE REVIEW

File: store/tikv/snapshot.go [hunk 1]
```diff
@@ -46,116 +48,108 @@ func newTiKVSnapshot(store *tikvStore, ver kv.Version) *tikvSnapshot {
...
+	var mu sync.Mutex
+	m := make(map[string][]byte)
+	err := s.batchGetKeysByRegions(bytesKeys, func(k, v []byte) {
+		if len(v) == 0 {
+			return
 		}
-		cmdBatchGetReq := singleBatchGet.CmdBatchGetRequest
-		cmdBatchGetReq.Keys = append(cmdBatchGetReq.Keys, k)
+		mu.Lock()
+		defer mu.Unlock()
```
Comment (ngaut) [inline store/tikv/snapshot.go]: Don't use defer here, there could be performance issue.

File: store/tikv/snapshot.go [hunk 2]
```diff
@@ -46,116 +48,108 @@ func newTiKVSnapshot(store *tikvStore, ver kv.Version) *tikvSnapshot {
...
+		}
+		var lockedKeys [][]byte
+		for _, pair := range batchGetResp.Pairs {
+			keyErr := pair.GetError()
+			if keyErr == nil {
+				collectF(pair.GetKey(), pair.GetValue())
+				continue
 			}
-			m, err = mergeResult(m, res)
+			val, err := s.handleKeyError(keyErr)
```
Comment (shenli) [inline store/tikv/snapshot.go]: Add a TODO here. If we meet many locks, this could be slow.

File: store/tikv/txn_committer.go [hunk 3]
```diff
@@ -339,7 +339,7 @@ type batchKeys struct {
 
 // appendBatchBySize appends keys to []batchKeys. It may split the keys to make
 // sure each batch's size does not exceed the limit.
-func appendBatchBySize(b []batchKeys, region RegionVerID, keys [][]byte, sizeFn func([]byte) int) []batchKeys {
+func appendBatchBySize(b []batchKeys, region RegionVerID, keys [][]byte, sizeFn func([]byte) int, limit int) []batchKeys {
```
Comment (coocood) [inline store/tikv/txn_committer.go]: limit is not used.

**Comment by shenli** [comment]
LGTM

File: store/tikv/txn_committer.go (hunk 3 above)
Comment (disksing) [inline store/tikv/txn_committer.go]: @coocood Forgot to update function body... PTAL.

**Comment by coocood** [comment]
LGTM

**Pull request merged**